compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -690,9 +701,36 @@ where
                                 tcx.mk_array(Option<fn()>),
                             ])
                             */
+                        } else {
+                            metadata
                         }
-                        _ => bug!("TyAndLayout::field({:?}): not applicable", this),
-                    }
+                    } else {
+                        match tcx.struct_tail_erasing_lifetimes(pointee, cx.param_env()).kind() {
+                            ty::Slice(_) | ty::Str => tcx.types.usize,
+                            ty::Dynamic(_, _, ty::Dyn) => {
+                                tcx.mk_imm_ref(
+                                    tcx.lifetimes.re_static,
+                                    tcx.mk_array(tcx.types.usize, 3),
+                                )
+                                /* FIXME: use actual fn pointers
+                                Warning: naively computing the number of entries in the
+                                vtable by counting the methods on the trait + methods on
+                                all parent traits does not work, because some methods can
+                                be not object safe and thus excluded from the vtable.
+                                Increase this counter if you tried to implement this but
+                                failed to do it without duplicating a lot of code from
+                                other places in the compiler: 2
+                                tcx.mk_tup(&[
+                                    tcx.mk_array(tcx.types.usize, 3),
+                                    tcx.mk_array(Option<fn()>),
+                                ])
+                                */
```

Comment:
Nit: can we extract the two copies of this code into a local function or closure? Duplicating the long comment, including the failed attempt counter, doesn't seem great to me 🙂